Move integration test into examples

mrchtr opened the pull request:
- moved integration test into examples `folder`
- adjust sample implementation to new interface and how a user would do it 
- added script to execute the pipeline using fondant cli
- add execution to ci/cd pipeline 

For now I've used a bash script for the execution. If we want to add more integration test we should think about a different approach. 

Fixes #727

RobbeSneyders commented on .github/workflows/pipeline.yaml:
```diff
@@ -32,6 +32,23 @@ jobs:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           COVERALLS_FLAG_NAME: test-${{ matrix.python-version }}
 
+  integration-test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python 3.9
+        uses: actions/setup-python@v1
+        with:
+          python-version: 3.9
+      - name: Install dependencies
+        run: |
+          pip install --upgrade pip
+          pip install poetry==1.4.0
+          poetry install --all-extras --with test
+
+      - name: Execute sample pipeline
+        run: ./scripts/run_sample_pipeline.sh
+
```
Would be nice to include this in `tox` as well.

RobbeSneyders reviewed (commented):
Thanks @mrchtr!

RobbeSneyders commented on scripts/run_sample_pipeline.sh:
```diff
@@ -0,0 +1,34 @@
+  --extra-volumes $resolved_path:/data
+
+# Expect that .artifacts was created and isn't empty
+if [ -d "./examples/sample_pipeline_test/.artifacts" ]; then
```
I guess this check is no longer needed now that you create the directory in this script?

RobbeSneyders commented on scripts/run_sample_pipeline.sh:
```diff
@@ -0,0 +1,18 @@
+    rm -r "$artifact_directory"
+    echo "Temporary directory removed"
+  fi
+}
```
I still think this is needed.

You might even want to catch the original exit code and reraise it like this:

```bash
cleanup() {
    rv=$?
    ...
    exit $rv
}
```

mrchtr replied:
I wasn't able to get the exit code of the failed docker container execution. I only get the exit code of the docker compose process. In particular, if a component fails with 1, and the compose succeeded with 0, I will miss the 1 since the docker compose process was completed successfully.

mrchtr replied:
I've added it. Now we catch exit codes that are not coming from the component code. Might be fine here, cause we are testing not the component code itself.

RobbeSneyders replied:
What exit codes are we catching then? Because we do want to catch errors from eg. the executor, which is also running in the container.

RobbeSneyders replied:
If docker compose is always exiting with code 0, we might want to look into the `--abort-on-container-exit` or `--exit-code-from` [options](https://docs.docker.com/engine/reference/commandline/compose_up/#options).

RobbeSneyders commented:
As discussed, we need to check each step in the propagation chain of the exit codes:
- Docker container to docker compose: Check current behavior and `--abort-on-container-exit` flag
- Docker compose to Fondant: We should at least use `check_call` when running docker compose [here](https://github.com/ml6team/fondant/blob/665805d46d8c3d94de70764d1d4726ce734093bd/src/fondant/pipeline/runner.py#L44)
- Fondant to test script: I think we need to reraise the exit code as mentioned in my comment above

mrchtr commented:
I've updated the code. 

> Docker container to docker compose: Check current behavior and --abort-on-container-exit flag
I've checked it and it is not needed. When we use `check_call` instead of `call` in the local runners code, the exception is the same.

I've updated the test script. The sample pipeline is located in `examples/sample_pipeline`. I've adjusted the code of the pipeline that is behaves as a user would expect it. You can go into the folder and run `fondant run local pipeline.py` for quickly test the pipeline. 

Beside that in the folder is a `run.sh` now. Which is used to execute the sample pipeline within the ci/cd pipeline. It exits with the exit code of the pipeline execution. 

I've adopted the ci/cd script. Now it scans the example folders and looks for `run.sh` scripts, execute each of them and tracks the exit codes. When one of the scripts failed, it will return the failed ones and exit with 1. Otherwise exit 0 when every test passed. 

This allows us to add more integration test samples to the example folder in the future.

mrchtr commented on .github/workflows/pipeline.yaml:
```diff
@@ -32,6 +32,26 @@ jobs:
+        run: |
+          pip install --upgrade pip
+          pip install poetry==1.4.0
+          poetry install --all-extras --with test
```
We should use the actual branch for the integration tests. This means we have to build the components first we want to test, to avoid versioning conflicts.

RobbeSneyders commented:
> * Docker container to docker compose: Check current behavior and `--abort-on-container-exit` flag

I think we are still missing this

mrchtr commented:
> > * Docker container to docker compose: Check current behavior and `--abort-on-container-exit` flag
> 
> I think we are still missing this

A bit of back and forth. In the meanwhile`--abort-on-container-exit` made it to main. I've merge latest main into this branch. Should be up to date now.

RobbeSneyders reviewed (approved):
Yeah I missed your comment about it not being needed before, my mistake. Looks good now.
